cotix/_universal_shape.py

```diff
@@ -0,0 +1,90 @@
+import equinox as eqx
+import jax
+from jax import numpy as jnp, tree_util as jtu
+from jaxtyping import Array, Float
+
+from ._abstract_shapes import AbstractShape, SupportFn
+from ._collisions import check_for_collision_convex, compute_penetration_vector_convex
+from ._geometry_utils import HomogenuousTransformer
+
+
+class UniversalShape(eqx.Module, strict=True):
+    parts: list[AbstractShape]
+    _transformer: HomogenuousTransformer
```

Comment:
do we still need to keep the position and velocity of the body if we can represent them with this transform? or do we mostly have this transform as identity? 
the reason I want to verify is that it might be easy to forget one or the other when we want to get the actually position of the body's shapes

Reply:
the problem with not keeping actual coordinates is that the rotations 'might' be painful. Though..... Now that I think about it, we can represent the velocity & the position as two matrices 3x3: 18 values instead of 6. But this would greatly simplify logic

Reply:
Also in terms of performance this might even be better, because of tensor cores... or might not

Reply:
Though, actually, the question is how to apply velocity: like, hm, I guess if we have
cos -sin x
sin cos y   =  P
0    0     1  

and to rotate it by dangle, and move it by dx/dy is:

cos(a+da) = cos(a)cos(da)-sin(a)sin(da)
sin(a+da)=sin(a)cos(da)+cos(a)sin(da)

hence if the delta matrix is:
cosda -sinda dx
sinda cosda dy = V
0        0           1

the product will be
P @ V = 
cos cosda - sin sinda | -cos sinda - sin cos da | cos dx - sin da + x 

does not really work neatly, so idk
like, we will have to manually overwrite the last coordinate i guess?

Reply:
> we can represent the velocity & the position as two matrices 3x3

didn't follow that
The position can be encoded in the HomogenuousTransformer, and I think velocity we can keep as is  - vector size 2

Reply:
but that is still kinda annoying, like, I want a single operation that does exactly what I want. But it looks like there is no such operation, simply because to rotate the body around its axis we need to shift it back to origin, rotate it, and then shift back to the original location

So it would be some annoying transformation like, we assign last columns to zero, multiply by rotation matrix, and then assign new last columns to position + velocity

Reply:
While it is still kiiinda shorter, it is not that much shorter I think. Also when doing any manipulations, like physics, you won't have a way to extract angles - if you will need them (which is unlikely, but possible)

Reply:
maybe we can pack the (x,y,theta) and velocities (dx,dy,dtheta) into one big vector, but then we will still have to construct matrices. Though, actually, when I think about it, maybe we can construct these matrices right away, and don't really care about caching: obviously, forward matrix is easily constructed, and backward matrix I guess can also be analytically computed, just I will have to apply a bit of thought.

Reply:
alright you certainly put more thought into it than I did. i think we can keep it like this. thanks for the info

Also, sorry, I just realized I said "position and velocity of the body" whereas I wanted to say "position and angle"

Reply:
ahh alright. Yeah, then your point makes sense, but the amount of pain that we will have to experience to apply velocity in this format might make this not worthwhile.

I am going to readup about applying velocity in homogenuous coordinates

Reply:
Alright, what do you think about representing the (x,y,theta) as a single vector of length 3, instead of (x, y) and (theta)? Like, it does make sense from some point of view, and doesn't from others. I cannot really decide

Reply:
Well, anyway, let's open issue about it, and continue discussion there. It is not hard to change HomogenuousTransformer into whatever else format we want